src/Microsoft.AspNetCore.SignalR.Common/Internal/Encoders/IDataEncoder.cs

```diff
@@ -1,11 +1,13 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+
 namespace Microsoft.AspNetCore.SignalR.Internal.Encoders
 {
     public interface IDataEncoder
```

Comment:
@davidfowl, @ahsonkhan, should this interface follow the IBufferOperation pattern? i.e. both input and output are in parameters. As designed today, the interface pretty much torpedoes all non allocating APIs. In case of pass through decoder, nothing is allocated, but the moment the decoder does something useful, it will have to allocate.

Reply:
@KrzysztofCwalina - we need to look at at the entire parsing/writing stack and remove byte[] there. I think it will be required when switching to pipelines.